pkg/services/ngalert/remote/client/alertmanager.go

```diff
@@ -0,0 +1,113 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/grafana/grafana/pkg/infra/log"
+	amclient "github.com/prometheus/alertmanager/api/v2/client"
+)
+
+const alertmanagerAPIMountPath = "/alertmanager"
+const alertmanagerReadyPath = "/-/ready"
+
+type AlertmanagerConfig struct {
+	TenantID string
+	Password string
+	URL      *url.URL
+	Logger   log.Logger
+}
+
+type Alertmanager struct {
+	*amclient.AlertmanagerAPI
+	httpClient *http.Client
+	url        *url.URL
+	logger     log.Logger
+}
+
+func NewAlertmanager(cfg *AlertmanagerConfig) (*Alertmanager, error) {
+	// First, add the authentication middleware.
+	c := &http.Client{Transport: &MimirAuthRoundTripper{
+		TenantID: cfg.TenantID,
+		Password: cfg.Password,
+		Next:     http.DefaultTransport,
+	}}
+
+	apiEndpoint := *cfg.URL
+
+	// Next, make sure you set the right path.
+	u := apiEndpoint.JoinPath(alertmanagerAPIMountPath, amclient.DefaultBasePath)
+	transport := httptransport.NewWithClient(u.Host, u.Path, []string{u.Scheme}, c)
+
+	return &Alertmanager{
+		logger:          cfg.Logger,
+		url:             cfg.URL,
+		AlertmanagerAPI: amclient.New(transport, nil),
+		httpClient:      c,
+	}, nil
+}
+
+// GetAuthedClient returns a *http.Client that includes a configured MimirAuthRoundTripper.
+// Requests using this client are fully authenticated.
+func (am *Alertmanager) GetAuthedClient() *http.Client {
+	return am.httpClient
+}
+
+// IsReadyWithBackoff executes a readiness check against the `/-/ready` Alertmanager endpoint.
+// If it takes more than 10s to get a response back - we abort the check.
+func (am *Alertmanager) IsReadyWithBackoff(ctx context.Context) (bool, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	attempt := func() (int, error) {
+		readyURL := am.url.JoinPath(am.url.Path, alertmanagerReadyPath)
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, readyURL.String(), nil)
+		if err != nil {
+			return 0, fmt.Errorf("error creating the readiness request: %w", err)
+		}
+
+		res, err := am.httpClient.Do(req)
+		if err != nil {
+			return 0, fmt.Errorf("error performing the readiness check: %w", err)
+		}
+
+		defer func() {
+			if err := res.Body.Close(); err != nil {
+				am.logger.Warn("Error closing response body", "err", err)
+			}
+		}()
+
+		return res.StatusCode, nil
+	}
+
+	var attempts int
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			attempts++
+			status, err := attempt()
+			if err != nil {
+				am.logger.Error("Ready check attempt failed", "attempt", attempts, "err", err)
+				continue
+			}
+
+			if status != http.StatusOK {
+				am.logger.Error("Ready check failed, status code is not 200", "attempt", attempts, "status", status, "err", err)
+				continue
+			}
+
+			return true, nil
+		case <-time.After(10 * time.Second):
```

Comment:
`time.After()` is creating being evaluated and creating a new channel on each loop iteration, we're retrying infinitely. We should create the timer outside of this loop:
```go
// Outside.
timeout := time.After(10 * time.Second)

// Inside
case <-timeout:
	cancel()
```


Since we're using a context for the readiness check, the easiest option would be to use a `context.WithTimeout(10 * time.Second)` instead of `context.WithCancel()`. We'd skip creating the timer and canceling the context on timeout.

Reply:
> time.After() is creating being evaluated and creating a new channel on each loop iteration, we're retrying infinitely. 

Thanks, good spot.

> Since we're using a context for the readiness check, the easiest option would be to use a context.WithTimeout(10 * time.Second) instead of context.WithCancel(). 

Semantically, I don't think this is correct even if they both achieve the same. `withTimeout` is typically used in the operation you're using the `ctx` on is meant to have a deadline. However, in our case the atomic operation (executing the HTTP request) is less important than the combination of them.

In essence, is different to say - I care about this HTTP call taking 10s vs I care about exceeding 10s while I make repeated HTTP calls. 

`WithTimeout` is commonly used on the former - so I'll stick with the current approach.

Reply:
Ah, nice to know! Go for the other approach then.